Automattic/wp-calypso
# Checkout: Remove undocumented.transactions function, take 2

This is a second attempt at https://github.com/Automattic/wp-calypso/pull/57280 which had to be reverted in https://github.com/Automattic/wp-calypso/pull/57504 due to an issue where the `address1` key in the contact details data was not being correctly converted to `address_1` before being sent to the endpoint. 

This PR greatly improves the `camelToSnakeCase` function to operate more like lodash's `snakeCase`. It includes unit tests for cases like the `address_1` issue above as well as many other cases that probably won't occur but cover other unusual rules used by `snakeCase`.

This PR also improves the unit tests for the payment processor functions so that they will also test for many of the expected contact details keys, including `address_1`.

Really, the contact details submission should be explicitly converted and typed as snake_case [like we do for contact validation](https://github.com/Automattic/wp-calypso/blob/cdce28a85cd69a0318e7d7d166b554487bfa8735/client/my-sites/checkout/composite-checkout/lib/contact-validation.tsx#L213), but this would be a much larger change due to the number of places that use the camelCase keys, so I think it's best to put that off until a later PR.

#### Changes proposed in this Pull Request

As part of the goal to remove the `wpcom.undocumented` module (see 429-gh-Automattic/payments-shilling), this PR modifies checkout to call the `/me/transactions` endpoint directly. This is the endpoint used to submit a purchase by all payment methods except for PayPal.

Since PayPal uses its own endpoint, it will need to be updated separately.

(This PR isn't as big as it looks. Mostly it's just updating the tests to mock the actual endpoint calls instead of being able to mock the `undocumented` function.)

#### Testing instructions

- Add a plan _and a domain_ to your cart and visit checkout.
- Complete the checkout form.
- Chose any payment method other than PayPal and submit the purchase.
- Verify that the purchase is successful.
